src/vcSHP.cpp

```diff
@@ -0,0 +1,741 @@
+  if (header.fileLength < UINT_MAX / 2)
+    header.fileLength *= 2;
+  else
+    header.fileLength = (UINT_MAX / 2) * 2;
+
+  header.version = (headerBuffer[31] << 24) | (headerBuffer[30] << 16) | (headerBuffer[29] << 8) | headerBuffer[28];
+  header.shapeType = (vcSHP_Type)((int32_t)((headerBuffer[35] << 24) | (headerBuffer[34] << 16) | (headerBuffer[33] << 8) | headerBuffer[32]));
+
+  offset = 36;
+  offset = UnpackDouble(pSHP->localBigEndian, offset, &header.MBRmin.x, headerBuffer);
+  offset = UnpackDouble(pSHP->localBigEndian, offset, &header.MBRmin.y, headerBuffer);
+  offset = UnpackDouble(pSHP->localBigEndian, offset, &header.MBRmax.x, headerBuffer);
+  offset = UnpackDouble(pSHP->localBigEndian, offset, &header.MBRmax.y, headerBuffer);
+  offset = UnpackDouble(pSHP->localBigEndian, offset, &header.ZRange.x, headerBuffer);
+  offset = UnpackDouble(pSHP->localBigEndian, offset, &header.ZRange.y, headerBuffer);
+  offset = UnpackDouble(pSHP->localBigEndian, offset, &header.MRange.x, headerBuffer);
+  offset = UnpackDouble(pSHP->localBigEndian, offset, &header.MRange.y, headerBuffer);
+
+  UD_ERROR_IF(offset != 100, udR_ParseError);
+  udFree(headerBuffer);
+
+  UD_ERROR_CHECK(pSHP->shpRecords.Init(32));
+  while (offset < header.fileLength)
+  {
+    result = vcSHP_LoadShpRecord(pSHP, pFile, offset, header.shapeType);
+    UD_ERROR_CHECK(result);
+  }
+
+epilogue:
+  udFile_Close(&pFile);
+
+  if (headerBuffer)
+    udFree(headerBuffer);
+
+  return result;
+}
+
+void vcSHP_LoadDbfFile(vcSHP *pSHP, const char *pFilename)
+{
+  vcDBF_Create(&pSHP->pDBF);
+  vcDBF_Load(&pSHP->pDBF, pFilename);
+}
+
+udResult vcSHP_LoadFileGroup(vcSHP *pSHP, const char *pFilename)
+{
+  if (pSHP == nullptr || pFilename == nullptr)
+    return udR_InvalidParameter_;
+
+  int i = 1;
+  if (*((uint8_t *)&i) == 1)
+    pSHP->localBigEndian = false;
+  else
+    pSHP->localBigEndian = true;
+
+  // load shp file
+  udResult result;
+  result = vcSHP_LoadShpFile(pSHP, pFilename);
+  if (result != udR_Success)
+  {
+    vcSHP_Release(pSHP);
+    return result;
+  }
+
+  printf("shape file load successfully %s\n", pFilename);
+
+  char extraFileName[256] = "";
+  int32_t nameLen = (int32_t)strlen(pFilename) - 3;
+  memcpy(extraFileName, pFilename, nameLen);
+  memcpy(extraFileName + nameLen, "dbf", 3);
+  vcSHP_LoadDbfFile(pSHP, extraFileName);
+
+  memcpy(extraFileName + nameLen, "prj", 3);
+  pSHP->projectionLoad = udFile_Load(extraFileName, &pSHP->WKTString);
+
+  return result;
+}
+
+void vcUDP_AddModel(vcState *pProgramState, udProjectNode *pParentNode, vcSHP_Record &record, vcDBF_Record *pDBFRecord, int32_t stringIndex)
+{
+  char buffer[256] = {};
+  const char *nodeName = "";
+  if (pDBFRecord != nullptr && !pDBFRecord->deleted && stringIndex >= 0)
+  {
+    size_t len = strlen(pDBFRecord->pFields[stringIndex].pString);
+    memcpy(buffer, pDBFRecord->pFields[stringIndex].pString, len);
```

Comment:
I'm pretty sure you can just use `udStrcpy` or `udStrncpy` instead of this, additionally, you should use `udStrlen` instead of `strlen`.

Reply:
fixed.